core/src/test/scala/pureconfig/ConfigReaderSuite.scala

```diff
@@ -0,0 +1,70 @@
+package pureconfig
+
+import com.typesafe.config.{ ConfigFactory, ConfigValue, ConfigValueFactory }
+import org.scalacheck.{ Arbitrary, Gen }
+import pureconfig.error.ConfigReaderFailures
+
+class ConfigReaderSuite extends BaseSuite {
+  val intReader = ConfigReader[Int]
+  val strReader = ConfigReader[String]
+
+  def intSummedReader(n: Int) = new ConfigReader[Int] {
+    def from(config: ConfigValue) = intReader.from(config).right.map(_ + n)
+  }
+
+  // generate configs that always read correctly as strings, but not always as integers
+  val genConfig: Gen[ConfigValue] =
+    Gen.frequency(95 -> Gen.chooseNum(Int.MinValue, Int.MaxValue), 5 -> Gen.alphaStr)
+      .map(ConfigValueFactory.fromAnyRef)
+
+  implicit val arbConfig = Arbitrary(genConfig)
+
+  behavior of "ConfigReader"
+
+  it should "have a correct map method" in forAll { (conf: ConfigValue, f: Int => String) =>
+    intReader.map(f).from(conf) === intReader.from(conf).right.map(f)
+  }
+
+  it should "have a correct flatMap method" in forAll { conf: ConfigValue =>
+    val g = { n: Int => intSummedReader(n) }
+    intReader.flatMap(g).from(conf) === intReader.from(conf).right.flatMap(g(_).from(conf))
+  }
+
+  it should "have a correct zip method" in forAll { conf: ConfigValue =>
+    def zip[A, B](r1: ConfigReader[A], r2: ConfigReader[B]): Either[ConfigReaderFailures, (A, B)] = {
+      (r1.from(conf), r2.from(conf)) match {
+        case (Right(a), Right(b)) => Right((a, b))
+        case (Left(fa), _) => Left(fa)
+        case (_, Left(fb)) => Left(fb)
```

Comment:
Right now, `zip` short-circuits in case the first reader fails. We could also use always both readers and combine their failures if both fail. What do you prefer?